Src/Newtonsoft.Json.UnityConverters/PartialConverter/ValuesArray.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Newtonsoft.Json.UnityConverters
+{
+    public readonly struct ValuesArray<TInner> : IReadOnlyList<TInner>, IEquatable<ValuesArray<TInner>>
+    {
+        private readonly TInner[] _array;
+
+        public int Count => _array.Length;
+
+        [MaybeNull]
+        public TInner this[int index]
+        {
+            get => _array[index];
+            set => _array[index] = value;
+        }
+
+        public ValuesArray(TInner[] array)
+        {
+            _array = array;
+        }
+
+        public ValuesArray(int capacity)
+        {
+            _array = new TInner[capacity];
+        }
+
+        [return: MaybeNull]
+        public TInner Get(int index)
+        {
+            return _array[index];
+        }
+
+        [return: MaybeNull]
+        public T GetAsTypeOrDefault<T>(int index)
+            where T : struct
+        {
+            return _array[index] as T? ?? default;
```

Comment:
Possible extra cast. Check if it's the correct type first before checking if it's the nullable variant?

Reply:
Ignoring. Wrote as XML comments how it's supposed to be used.